.github/workflows/ci.yml

```diff
@@ -11,6 +11,10 @@ env:
   CARGO_TERM_COLOR: always
   NIGHTLY_TOOLCHAIN: nightly
 
+concurrency:
+  group: ${{github.workflow}}-${{github.ref}}
```

Comment:
@lee-orr this workflow runs on pushes to `main` (as do the workflows below). When that event occurs, `github.ref` is `refs/head/main` (see https://docs.github.com/en/actions/learn-github-actions/contexts). This means that multiple pushes to `main` causes this to cancel existing runs. I'm not sure if that was intended in this PR, or if you only intended to cancel existing PR runs?

If the intent is to let all `main` workflows run and not be cancelled, I think we might need to use the approach taken here, to reference a codebase in the wild: https://github.com/tailscale/tailscale/blob/6540d1f01800d03fa46d467fae777163b8ba3d0d/.github/workflows/test.yml#L30-L38

i.e. set `group` to `${{ github.workflow }}-$${{ github.pull_request.number || github.run_id }}`.

I refer to the comment in the linked YAML which explains it perfectly:

> For PRs, later CI runs preempt previous ones. e.g. a force push on a PR
> cancels running CI jobs and starts all new ones.
> 
> For non-PR pushes, concurrency.group needs to be unique for every distinct
> CI run we want to have happen. Use run_id, which in practice means all
> non-PR CI runs will be allowed to run without preempting each other.

Reply:
Edit: I took another look and realised that `github.pull_request.number` doesn't seem to be documented in https://docs.github.com/en/actions/learn-github-actions/contexts. It could be undocumented or the repo I've used as a reference isn't working as expected...

In any case, I believe the idea is still correct, but instead we want to use the data off the `github.event` variable.

i.e. set `group` to `${{ github.workflow }}-$${{ github.event.pull_request.number || github.run_id }}`.

See the `pull_request` event payload fields https://docs.github.com/en/webhooks/webhook-events-and-payloads#pull_request. Note: You could use `github.event.number` but the one under `pull_request` seems more explicit of the intent.

Reply:
@mgi388 - my original intent was that it also cancel main runs. Essentially - any branch or PR can only have a single active workflow at the time, on the most recent commit - if a new commit is introduced it will be replaced. My reasoning was that we only really need to know what the state of the current codebase or PR is. Is there any reason we should not cancel runs on main when they are already outdated?

Reply:
Being able to isolate exactly which commit broke main is really helpful.

Reply:
Makes sense, I was under the impression that the merge queue system already breaks that capacity by collecting a bunch of PRs into a single commit instance and running tests on groups of PRs? If that's not the case, then yes - we can make an adjustment so main doesn't self-cancel.

Reply:
Probably already what @alice-i-cecile is saying, but yeah I believe that if you cancel previous runs, that means you'll look through the commit history and instead of green success ticks, you'll see that it was cancelled so either grey or red. E.g.

<img width="682" alt="image" src="https://github.com/bevyengine/bevy/assets/135186256/1af1b8be-6f7f-49ff-b2ba-54eea8df8e48">

Reply:
It breaks it a bit, but we want all the granularity we can still.

Reply:
Sounds good - will adjust